nixpkgs-manual-mmdoc, nixos-manual-mmdoc: init

ryantm opened the pull request:
This is a prototype Markdown manual renderer that has as few dependencies as possible. We might wish to have something like this because it would limit the build closure required for building the manuals, which is most important for NixOS.

The only dependencies are:
* [mmdoc](https://github.com/ryantm/mmdoc) - My C wrapper around cmark-gfm for making documentation
* [cmark-gfm](https://github.com/github/cmark-gfm) - a C implementation of the CommonMark spec with additional GitHub-flavored Markdown extensions
* xxd
* fastJson
* libzip

It uses assets from
* fuse.js
* highlight.js

Another benefit of this minimal approach is the manual can be rendered very quickly. cmark claims it can render War and Peace in 100 ms. 

## [Try the Live demo!](http://ryantm.github.io/nixpkgs/)

## Try it yourself

* Checkout this branch
* nix run .#nixpkgs-manual-mmdoc-watch
* visit `http://localhost:8000/` in your web browser

jtojnar commented on doc/languages-frameworks/index.md:
```diff
@@ -0,0 +1,31 @@
+
+The standard build environment makes it easy to build typical Autotools-based packages with very little code. Any other kind of package can be accomodated by overriding the appropriate phases of `stdenv`. However, there are specialised functions in Nixpkgs to easily build packages for other programming languages, such as Perl or Haskell. These are described in this chapter.
+
+* [Agda](agda.html)
```
Hardest part are the anchors. You will want something independent from output format and decoupled from file structure. That way you can render it to anything, including a single-page HTML. This is where document-oriented tools like Docbook with document-wide identifiers shine and tools that are file-based have to resort to ugly hacks.

ryantm replied:
Okay, I'll tackle this issue next.

ryantm replied:
I have a new anchor system now. There is a anchors.nix where you associate an anchor with a file. Right now it only supports HTML but I'll update that to be more generic.

ryantm commented:
I've added a table of contents now.

![image](https://user-images.githubusercontent.com/4804/103425380-d65b7f00-4b66-11eb-9a16-076e9b7f211f.png)

Mic92 commented on doc/minimal-manual.nix:
```diff
@@ -0,0 +1,41 @@
+      outputFile="''${outputFile%.chapter}"
+      outputFile="$out/''${outputFile#"$src/"}.html"
+      mkdir -p "$(dirname "$outputFile")"
+      echo "${htmlHead}" >> "$outputFile"
+      if [ $file != "$src/toc.md" ]; then
+        echo "$toc" >> "$outputFile"
+      fi
+      echo "<section>" >> "$outputFile"
+      cat "$file" | ${cmd} >> "$outputFile"
+      echo "</section>" >> "$outputFile"
+      echo "${htmlFoot}" >> "$outputFile"
```
some here-doc would be more readable here.

ryantm replied:
Yep, it did, thanks.

ryantm commented:
Now with syntax highlighting!

![image](https://user-images.githubusercontent.com/4804/103430511-baada400-4b79-11eb-9083-494c889c3408.png)

jtojnar commented on doc/anchors.nix:
```diff
@@ -0,0 +1,32 @@
+{
```
Could this file be auto-generated? And conflicts checked?

ryantm replied:
I think so!

jtojnar replied:
Also do not forget about the [bracketed spans](https://pandoc.org/MANUAL.html#extension-bracketed_spans) (used in GNOME docs). Unfortunately, those are not expressible in regex very nicely.

ryantm replied:
It might be sufficient to look for all the instances of `{#sec-name}`.

ryantm replied:
Auto generated anchor rewriting, and bracketed span rewriting is working!

jtojnar commented on doc/minimal-manual.nix:
```diff
@@ -0,0 +1,75 @@
+  sedRewritesPost = joinRewrites rewriteAnchors;
+
+  sedCmdPre = "${gnused}/bin/sed -r '${sedRewritesPre}'";
+  cmarkCmd = "${cmark-gfm}/bin/cmark-gfm --unsafe";
```
Could you add a comment why unsafe is needed?

ryantm replied:
If we do not use `--unsafe`, all the raw HTML gets output as:
```
<!-- raw HTML omitted -->
```

Currently, I'm using raw HTML for the admonition and the header ids.

jtojnar commented on doc/minimal-manual.nix:
```diff
@@ -0,0 +1,95 @@
+  rootPath = "http://localhost:8000/result";
+
+  rewriteHeader = ''s|# ([^{]*)\s\{(#([^}]*))}$|# <span id="\3">[\1](\2)</span>|g'';
+  rewriteAdmonitionStart = ''s|::: warning|<div class="warning"><h3 class="title">Warning</h3><p>\n|g'';
```
I think we use note admonition as well at least, here is a full list:

https://github.com/jgm/pandoc/pull/6922/files#diff-d3c4d5818e408935dae861295b6e9b73434ea63069e7b08d2c06ece876f271a7R198

ryantm replied:
added support for all the admonition types.

ryantm commented:
I think I have everything that is converted in the table of contents now:

![image](https://user-images.githubusercontent.com/4804/103433974-0cbcec80-4baf-11eb-9f28-aba07177b22a.png)

ryantm commented:
Currently, there is no support for dd/dl lists like are being used in Cross-compilation. I kind of don't want to add those since it seems hard.

ryantm reviewed (commented):
~~Some of the edits here should be removed and it should be rebased on top of #108147~~

ryantm commented:
Now it has a really basic search, see screenshot in top post.

ryantm commented:
I've extracted the work into its own project now https://github.com/ryantm/mmdoc

Mic92 commented on pkgs/tools/text/mmdoc/default.nix:
```diff
@@ -0,0 +1,67 @@
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p "$out/bin"
+    mkdir -p "$out/lib/mmdoc"
+    install "$src/bin/mmdoc" "$out/bin"
+    install "$src/lib/mmdoc/fuse.basic.min.js" "$out/lib/mmdoc"
+    install "$src/lib/mmdoc/highlight.pack.js" "$out/lib/mmdoc"
+    install "$src/lib/mmdoc/minimal.css" "$out/lib/mmdoc"
+    install "$src/lib/mmdoc/mono-blue.css" "$out/lib/mmdoc"
+    install "$src/lib/mmdoc/search.js" "$out/lib/mmdoc"
```
Just a bit shorter and less repetition:

```suggestion
    install -D "$src/bin/mmdoc" "$out/bin/mmdoc"
    install -D -t "$out/lib/mmdoc" \
       "$src/lib/mmdoc/"{fuse.basic.min.js,highlight.pack.js,minimal.css,mono-blue.css,search.js}
```

ryantm replied:
I should probably move this to some Makefile in the project once I learn how.

ryantm replied:
Moved it to a Makefile and used your 2-liner. https://github.com/ryantm/mmdoc/blob/main/Makefile thanks!

Mic92 commented on pkgs/tools/text/mmdoc/default.nix:
```diff
@@ -0,0 +1,67 @@
+    install "$src/lib/mmdoc/minimal.css" "$out/lib/mmdoc"
+    install "$src/lib/mmdoc/mono-blue.css" "$out/lib/mmdoc"
+    install "$src/lib/mmdoc/search.js" "$out/lib/mmdoc"
+    patchShebangs $out/bin
```
I think stdenv does this in fixup phase?

ryantm replied:
It doesn't work if I don't do that though.

jtojnar replied:
That's because you are overriding `fixupPhase`. See https://github.com/jtojnar/nixpkgs-hammering/blob/e414d83e8fbbc920674ee8ae239395704f3b5d54/explanations/fixup-phase.md

ryantm replied:
Fixed all nixpkgs-hammering issues!

Mic92 commented:
As this would be a critical part of nixpkgs/NixOS it might make sense to move it over to the nix-community or NixOS org later.

ryantm commented:
> As this would be a critical part of nixpkgs/NixOS it might make sense to move it over to the nix-community or NixOS org later.

Sounds fine. I extracted it when I realized it might be generally useful for other projects. I'm experimenting with using it for my personal website. 

It may make sense to vendor it into nixpkgs like we do with highlight.js so people can build the manual offline more reliably. I'm asduming that's why we vendor it that way.

ryantm commented:
There is now support for single page mode. The way it works is that it uses the order anchor references appear in toc.md to add the sections to the page.

Roadmap going forward:

- [ ] Man pages
- [ ] Better styling
- [ ] Expose all three types ( single, multi, manpage) manuals as attrpaths that can be built

samueldr commented:
>  Better styling

Try to keep this flexible enough so that the default styles of your project don't direct too much of the DOM structure.

(I'm keeping tabs on this since it maybe could end-up being used, and if so, would be used for the website docs I figure.)

ryantm commented:
I wasn't happy with mmdoc being implemented in Bash so I'm [rewriting it in C now](https://github.com/ryantm/mmdoc/tree/c).

jtojnar commented:
Does it have to be that minimal? Won’t Python or Perl or at least C++ deps be in the closure anyway?

flokli commented:
One could even argue "do we need to write our own" as well. If we're fine depending on a compiler, maybe there's something a bit less heavy than pandoc already written.

samueldr commented:
I'm still interested in having more options open for us. First here, I can see this is Ryan's project for Ryan. Then, it can be a good option if it's slim and quick enough.

I don't like the whole concept of "we already have _that bloat_ on an installed system, so we should further increase our dependency on _that bloat_." It makes it harder to remove or reduce later on.

(Though, I do understand and appreciate that you're trying to prevent needless NIH work :))

ryantm commented:
My main reason for working on it is I don't see other low-dependency/closure size alternatives, and it seems this is a requirement for using Markdown in the NixOS manual. I spent a few hours looking for one and didn't see anything.

ryantm commented:
I'm not sure if I'm misunderstanding flokli's comment about compilers, but if we eventually adopt mmdoc, then we'd be depending on a C compiler by having cmark-gfm as a dependency, so using C for mmdoc shouldn't add anything.

flokli commented:
My comment "If we're fine depending on a compiler" was a bit ironic - we won't get around to not having gcc in the build closure, as well as some pretty commonly used programming languages and build tools.

Using something like ruby, nodejs or haskell is obviously too big due to its build closures, but maybe there's a compromise. While I appreciate all the work that went into this, I'm indeed a bit afraid of creating an entire tool *just to render nixpkgs documentation*.

I stumbled over https://github.com/fletcher/MultiMarkdown-6, which seems to be pretty much self-contained (except using `cmake` as a build system, but we need that anyways). It supports Math, TOC, Glossaries, Footnotes and a bunch of output formats, and is probably more sustainable in the longer run than writing our own.

Mic92 commented:
> Does it have to be that minimal? Won’t Python or Perl or at least C++ deps be in the closure anyway?

So far the executable is just a thin wrapper around `cmark-gfm`, which does not look too scary to me. From what I see is that ryantm's solution gives us a lot of freedom regarding styling (css & html), which is probably what we need.

ryantm commented:
MultiMarkdown-6 looks like it could be good, thanks for pointing it out! One issue I noticed with it is the way it requires the anchor text to match the section heading text, which means if we switch to it and every time we change a section name, it would break the links to that section.

flokli commented:
> MultiMarkdown-6 looks like it could be good, thanks for pointing it out! One issue I noticed with it is the way it requires the crossreference anchor text to match the section heading text, which means if we switch to it and every time we change a section name, it would break the links to that section.

We can include an optional label in the heading if we refer to thing, to make these less likely to change. I'm not sure how it handles dead links, whether it fails the build or will just point into the void.

ryantm commented:
I'm getting close to feature parity with my bash script for the single page output. I figured out how to use the AST API that libcmark provides to make simple transformations and used that to transform bracketed spans that are just ids.

It does not do admonitions yet. I think I'm much more confident that I could do things like the data definition lists too.

It compiles all the markdown in the nixpkgs manual into a single HTML page in 11 ms on my machine :speed

ryantm commented:
I've added admonition rewriting and a test suite.

Mic92 commented on doc/minimal-manual.nix:
```diff
@@ -9,5 +9,7 @@ stdenv.mkDerivation rec {
-  builder = writeScript "${name}-builder" "${mmdoc}/bin/mmdoc -s $src $out";
+  buildInputs = [gdb mmdoc];
+
+  builder = writeScript "${name}-builder" "${gdb}/bin/gdb ${mmdoc}/bin/mmdoc";
```
Just a reminder to remove this.

ryantm commented:
I updated this PR to use the C implementation of mmdoc and I updated the first post to reflect the changes.

ryantm commented:
The C implementation is at our beyond feature parity with the bash implementation now. It now outputs both single and multipage formats (instead of letting you choose which one). Searching works for the multipage one. I updated the top post again to reflect these changes.

ryantm commented:
Now it outputs man pages too, but I don't know much about man pages so I probably did something obviously wrong.

ryantm commented:
Got a rudimentary epub output working by just wrapping the single page html with the appropriate epub headers. cmark html output was designed to work with epub https://github.com/commonmark/cmark/issues/80

ryantm commented:
After much work, the epub output is somewhat functional: it loads it on my Kobo eReader, but the code block get cut off if they are too wide.

ryantm commented:
@samueldr I feel ready to start collaborating on the styling of the webpage outputs.

Maybe you can take a closer look at this PR and https://github.com/ryantm/mmdoc and tell me what you think?

Mic92 commented:
> After much work, the epub output is somewhat functional: it loads it on my Kobo eReader, but the code block get cut off if they are too wide.

Also I wonder if we really need to support this just because we can. It's a rather uncommon usecase for documentation that is not a longer tutorial.

ryantm commented:
> Also I wonder if we really need to support this just because we can. It's a rather uncommon usecase for documentation that is not a longer tutorial.
 
Yeah, maybe. I think the current manual supports EPUB, so I wanted to attempt it too. I think it is close being decent. I need to figure out how to wrap or shrink code blocks.

ryantm commented:
I've been working on the styling a lot. You see a live demo of the latest work by looking at the mmdoc documentation, which is rendered by mmdoc:

https://ryantm.github.io/mmdoc/

ryantm commented:
I just updated this to include all the new markdown pages, and the latest version of mmdoc. The first comment is updated to include a link to a live demo!

http://ryantm.github.io/nixpkgs/

zseri commented:
Is it normal that the syntax highlighting is so.. ehm... *minimal*?

tomberek commented:
beautiful, and i like the instant search. I did a quick test on the feasibility to use this for the Nix manual. Needed some slight modifications to get the TOC to link up everything, but seemed to be functional!

davidak commented:
This already looks very good!

>My main reason for working on it is I don't see other low-dependency/closure size alternatives, and it seems this is a requirement for using Markdown in the NixOS manual. I spent a few hours looking for one and didn't see anything.

Have you looked at general static-site-generators like Hugo? (Closure size ~90 MiB, https://hydra.nixos.org/build/140332734#tabs-details)

Have you thought about how the whole website could be built with this infrastructure? Would it make sense to have one build pipeline for the complete website including all manuals?

Do you have ideas for a design where the manuals are tightly integrated into the website, from an end-user perspective. The current website works well. To the user, it don't feels like a different site.

ryantm commented:
> Is it normal that the syntax highlighting is so.. ehm... _minimal_?

I should choose a different color scheme for highlight.js. I'm currently using the one used by the existing manual, but it only uses shades of blue which seems bad for contrast.

> Have you looked at general static-site-generators like Hugo?

Not really, no. I'm also not actually sure what is an acceptable closure size cutoff to be used for the NixOS manual.

> Have you thought about how the whole website could be built with this infrastructure?

No, but I expect mmdoc will be needlessly constraining to use for a general purpose website. I am working on using it for my personal website though. 

> Would it make sense to have one build pipeline for the complete website including all manuals?

I don't know how the website is built currently, but I don't think anything here limits it from being built alongside the rest of the website.

> Do you have ideas for a design where the manuals are tightly integrated into the website, from an end-user perspective.

Not yet. I'll probably need to add a bit of templating so some HTML and styling can be injected to style it seamlessly.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/rfc-0072-fcp-switch-to-commonmark-for-documentation/9560/6

ryantm commented:
Just made a new release of mmdoc with improved styling. Here's the parts of the nixpkgs manual written in CommonMark built with mmdoc: https://ryantm.github.io/nixpkgs/

asymmetric commented:
@ryantm anything in particular we should look out for?

ryantm commented:
@asymmetric Any feedback is welcome, I feel like I definitely need more feedback.

For the code highlighting, I'm using a style that is meant for accessibility, (maybe high contrast, I'm not sure). Maybe take a look at https://highlightjs.org/static/demo/ and see if you have a better suggestion. I'm using A11y light and dark.

Sub-heading anchors already exist, but I definitely should figure out how to expose all the anchors in the UI!

asymmetric commented:
>  I'm using A11y light and dark.

When I look at the demo for Markdown, the `code` highlight isn't highlighted in the same way (with the big grey box). Is it something you're toggling separately?

ryantm commented:
Oh, you are talking about the inline highlighting! No, there is no colors for that. I agree that it is too intense. I'll try tone it down.

asymmetric commented:
Nix syntax highlighting seems pretty bare-bones, e.g. [here](https://ryantm.githubio/nixpkgs/builders/packages/kakoune/#sec-kakoune) only `with` is highlighted.

zseri commented on doc/doc-support/lib-functions-docs-cm.nix:
```diff
@@ -0,0 +1,24 @@
+# Generates the documentation for library functons via nixdoc. To add
```
```suggestion
# Generates the documentation for library functions via nixdoc. To add
```

zseri commented on pkgs/tools/nix/nixdoc/nixdoc-commonmark.nix:
```diff
@@ -0,0 +1,25 @@
+    hash = "sha256-kG/pIhqIQYuPfKG5LXOTKjNVpNldvpoL0/6jptodn4U=";
+  };
+
+  buildInputs =  lib.optional stdenv.isDarwin [ darwin.Security ];
```
```suggestion
  buildInputs = lib.optional stdenv.isDarwin [ darwin.Security ];
```

zseri commented on doc/minimal-manual.nix:
```diff
@@ -11,5 +11,12 @@ stdenv.mkDerivation rec {
+  buildPhase = ''
+    cp -r $src source
+    chmod -R u+w source
+    cp ${import ./doc-support/lib-functions-docs-cm.nix {}}/*.md source/functions/library/
```
imo `pkgs` should be propagated
```suggestion
    cp ${import ./doc-support/lib-functions-docs-cm.nix { inherit pkgs; }}/*.md source/functions/library/
```

zseri commented on doc/doc-support/lib-functions-docs-cm.nix:
```diff
@@ -0,0 +1,24 @@
+# Generates the documentation for library functons via nixdoc. To add
+# another library function file to this list.
```
this sentence sounds incomplete...
> To add another library function file to this list.

do what then?

zseri commented on doc/minimal-manual.nix:
```diff
@@ -11,5 +11,12 @@ stdenv.mkDerivation rec {
+    cp -r $src source
+    chmod -R u+w source
+    cp ${import ./doc-support/lib-functions-docs-cm.nix {}}/*.md source/functions/library/
+    ${mmdoc}/bin/mmdoc nixpkgs source $out
```
is `buildInputs = [ mmdoc ];` even necessary when we refer to `mmdoc` by absolute path?

ryantm replied:
Thanks; fixed!

zseri commented on doc/minimal-manual.nix:
```diff
@@ -0,0 +1,20 @@
+
+with pkgs;
+
+stdenv.mkDerivation rec {
```
```suggestion
stdenvNoCC.mkDerivation {
```

zseri commented on pkgs/tools/text/mmdoc/default.nix:
```diff
@@ -0,0 +1,40 @@
+{lib,
+stdenv,
+fetchFromGitHub,
+
+cmark-gfm,
+xxd,
+fastJson,
+libzip,
+self,
+ninja,
+meson,
+pkg-config
+} :
```
```suggestion
{ lib
, stdenv
, fetchFromGitHub

, cmark-gfm
, fastJson
, libzip
, ninja
, meson
, pkg-config
, xxd
}:
```

ryantm commented:
* Got graphviz diagram of staging working
* Added lib.asserts and lib.attrsets auto generation (still need to finish up making lib.attrsets good though)

[Live demo still available](https://ryantm.github.io/nixpkgs/).

zseri commented on doc/contributing/staging_workflow.dot:
```diff
@@ -0,0 +1,17 @@
+  "broken staging-next fixes" -> "staging-next"
+
+  "staging-next" -> master [color="#E85EB0"] [label="stabilization ends"] [fontcolor="#E85EB0"]
+  "staging" -> "staging-next" [color="#E85EB0"] [label="stabilization starts"] [fontcolor="#E85EB0"]
```
```suggestion
  staging -> "staging-next" [color="#E85EB0"] [label="stabilization starts"] [fontcolor="#E85EB0"]
```

zseri commented on lib/attrsets.nix:
```diff
@@ -171,15 +187,12 @@ rec {
   */
   filterAttrsRecursive =
-    /*
-String -> Any -> Bool
```
type information is lost when this part of the comment is removed as of now.

ryantm replied:
I'm leaving that out on purpose because the formatting is bad. The type is still captured by the overall type signature.

davidak commented:
Looked at it again.

This is how it looks on mobile (Android Phone, Firefox Klar):

![signal-2021-06-19-150917](https://user-images.githubusercontent.com/91113/122644159-25e13780-d114-11eb-83a1-4a796ba0a64d.png)

can't even toggle the menu

This is how it looks on desktop (Chromium 91):

![Screenshot from 2021-06-19 15-33-20](https://user-images.githubusercontent.com/91113/122644194-4f01c800-d114-11eb-802c-02731f84913c.png)

Can you make the menu not like an overlay that is over the text, but a grid layout, so you can see both?

See how https://doc.rust-lang.org/book/ reacts when you make the window small.

When i click on a topic in the menu, the complete page reloads. I have to scroll down the menu again... (again, mdBook don't has this issue)

Can you make the dark theme a dark grey instead of complete black? It hurts my eyes. See the coal theme of rust manual.

It seem what you build is a mdBook clone, in C. So when we decide to build a new tool, it shouldn't be worse than existing tools we could just use.

davidak commented on doc/toc.md:
```diff
@@ -0,0 +1,99 @@
+  * [](#sec-pkgs-dockerTools)
+  * [](#sec-pkgs-ociTools)
+  * [](#sec-pkgs-snapTools)
+*[](#chap-language-support)
```
```suggestion
* [](#chap-language-support)
```

that should fix it i guess

ryantm replied:
thanks, should be fixed

ryantm commented:
Thanks for the feedback. Yes, still work needs to be done on the styling. I'm not an expert at CSS, so if someone else would offer a hand, I'd be appreciative. I will, of course, keep working on it the best I can.

> existing tools we could just use

For the NixOS manual, it is not clear to me that we can "just" use mdbook. Since the manual is always built, it is important to keep the build time fast and closure-size small, which I doubt mdbook has as a requirement.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/happy-docsgiving/10214/28

davidak commented:
FYI: GNOME has a new documentation site and i think the design is perfect.

https://developer.gnome.org/documentation/introduction/flatpak.html

ryantm commented:
@davidak Thanks for sharing! Nice, their CSS looks very minimal (and CC0 licensed!) too. I like how the contents are collapsible and how there is a heading outline.

SuperSandro2000 commented on pkgs/tools/nix/nixdoc/nixdoc-commonmark.nix:
```diff
@@ -0,0 +1,25 @@
+    hash = "sha256-m+Du8n3kDoez7xbTV6FGF+bgJ2tSTGg2DclYKlAyGxA=";
+  };
+
+  buildInputs = lib.optional stdenv.isDarwin [ darwin.Security ];
```
```suggestion
  buildInputs = lib.optional stdenv.isDarwin [ Security ];
```

SuperSandro2000 commented on pkgs/tools/text/mmdoc/default.nix:
```diff
@@ -0,0 +1,38 @@
+
+  nativeBuildInputs = [ ninja meson pkg-config ];
+
+  buildInputs = [ cmark-gfm fastJson xxd libzip.dev ];
```
```suggestion
  buildInputs = [ cmark-gfm fastJson xxd libzip ];
```

should be choosen automatically.

SuperSandro2000 commented on pkgs/tools/text/mmdoc/default.nix:
```diff
@@ -0,0 +1,38 @@
+    license = licenses.cc0;
+    maintainers = with maintainers; [ ryantm ];
+    platforms = platforms.unix;
+  };
+
+}
```
```suggestion
  };
}
```

SuperSandro2000 commented on pkgs/tools/text/mmdoc/default.nix:
```diff
@@ -0,0 +1,38 @@
+
+  meta = with lib; {
+    description = "Minimal Markdown Documentation";
+    homepage = src.meta.homepage;
```
```suggestion
    homepage = "https://github.com/ryantm/mmdoc";
```

otherwise I can't click on it.

SuperSandro2000 commented on pkgs/tools/nix/nixdoc/nixdoc-commonmark.nix:
```diff
@@ -0,0 +1,25 @@
+    homepage = "https://github.com/nix-community/nixdoc";
+    license = [ licenses.gpl3 ];
+    maintainers = [ maintainers.tazjin ];
+    platforms = platforms.unix;
```
```suggestion
```
default of buildRustPackage

SuperSandro2000 commented on doc/minimal-manual.nix:
```diff
@@ -0,0 +1,22 @@
+    chmod -R u+w source
+    cp ${import ./doc-support/lib-functions-docs-cm.nix { inherit pkgs; }}/*.md source/functions/library/
+    FONTCONFIG_FILE=${fontconfig.out}/etc/fonts/fonts.conf \
+      ${graphviz}/bin/dot -Tsvg source/contributing/staging_workflow.dot > source/staging_workflow.svg
+    ${mmdoc}/bin/mmdoc nixpkgs source $out
```
Why not use (native)buildInputs?

SuperSandro2000 commented on doc/minimal-manual.nix:
```diff
@@ -0,0 +1,22 @@
+
+  src = builtins.filterSource (path: type: type == "directory" || builtins.match ".*\.md" path == [] || builtins.match ".*\.dot" path == []) ./.;
+
+  phases = [ "buildPhase" ];
+
```
```suggestion
```

overwriting phases is an anti pattern.

06kellyjac commented on doc/doc-support/lib-functions-docs-cm.nix:
```diff
@@ -0,0 +1,25 @@
+# another library file here, you also need to add it to the `doc/toc.md`.
+{ pkgs ? import ../.. {} }:
+
+with pkgs; stdenv.mkDerivation {
```
Can this use the noCC stdenv?

06kellyjac commented on doc/doc-support/lib-functions-docs-cm.nix:
```diff
@@ -0,0 +1,25 @@
+  buildInputs = [ nixdoc-commonmark ];
+  installPhase = ''
+    function docgen {
+      nixdoc -c "$1" -d "$2" -f "../lib/$1.nix"  > "$out/$1.md"
```
```suggestion
      nixdoc -c "$1" -d "$2" -f "../lib/$1.nix" > "$out/$1.md"
```

06kellyjac commented on doc/minimal-manual.nix:
```diff
@@ -0,0 +1,22 @@
+
+with pkgs;
+
+stdenvNoCC.mkDerivation rec {
```
Is rec actually used here?

06kellyjac commented on doc/minimal-manual.nix:
```diff
@@ -0,0 +1,22 @@
+# Minimal manual prototype
+
+{ pkgs ? (import ../. {}), nixpkgs ? { }} :
```
Is `nixpkgs` used? I couldnt spot it

06kellyjac replied:
```suggestion
{ pkgs ? (import ../. { }), nixpkgs ? { }} :
```

raboof commented:
Looking pretty great! Are there any known limitations left with this PR? The asserts and attrset library docs aren't linking to their code on github anymore, but that seems like an acceptable limitation to me. Anything further that needs to be fixed, aside from the merge conflicts?

jtojnar commented:
We need a decision whether to go with this or https://github.com/NixOS/nixpkgs/pull/105036.

raboof commented:
Do we have an overview of what is still missing here? I noticed:
* rebase :)
* chapter/section numbers: I think it's really helpful to have them numbered, even if the numbers aren't stable long-term.
* search styling seems incomplete
* search doesn't seem to quite work: I get results that don't (directly?) contain the search term (though it even seems to fail a bit differently locally vs on http://ryantm.github.io/nixpkgs/ - I tried with search term 'meta')

ryantm commented:
@raboof I rebased this and added all the new sections to the table of contents.

## section numbering

I am not excited about adding this feature. On top of the concern you brought up of them being unstable, it requires the table of contents to be more structured, which is something I don't like. I'd rather it be more curated by people. One example of this issue is the automated table of contents in the current manual does not put Haskell at the top level, and this frustrated them enough to move their documentation out of the manual to a separate site.

## styling

There are various bugs with the styles in mmdoc. At my current CSS skill level, I probably need to invest a few weeks of my time to fix it, so it might be a long time before that happens. If someone more skilled at CSS who is also interested in writing very minimal CSS would like to help out, I'd be excited to show them what's going on!

## path forward?

Maybe a good path forward would be to try to modify this MR in such a way that it can be merged into nixpkgs without forcing the manual to switch over in a more real way. Like maybe we can merge in the .md auto generated docs first, then this can probably go in without being too disruptive on the current manual

ryantm commented:
As usual, the most recent preview of this PR's manual is at http://ryantm.github.io/nixpkgs/

ryantm commented:
@raboof I messed with the search more and learned it was only searching the first 80 characters of each page! I also turned the fuzzy match waaay down. Seems kind of reasonable now at https://ryantm.github.io/nixpkgs/builders/packages/linux/

asymmetric commented:
Another feature I would find pretty important are anchor links for sub-paragraphs, e.g. to link to "Attributes used by the builder" in the [Go section](https://ryantm.github.io/nixpkgs/languages-frameworks/go/#sec-language-go).

The old manual [has this](https://nixos.org/manual/nixpkgs/stable/#ssec-go-common-attributes) and IMO it would be a shame to lose it.

ryantm commented:
Yes, I totally want that @asymmetric. There is an open issue for it https://github.com/ryantm/mmdoc/issues/17

ryantm commented:
@asymmetric Okay, headers have links now (if an anchor was specified for it): https://ryantm.github.io/nixpkgs/languages-frameworks/go/#ssec-go-common-attributes

doronbehar commented:
@ryantm is there a plan to also convert NixOS' documentation to CommonMark? (https://ryantm.github.io/nixos/)

ncfavier commented on doc/doc-support/lib-functions-docs-cm.nix:
```diff
@@ -0,0 +1,25 @@
+
+    mkdir -p $out
+
+    docgen asserts 'Assert functions'
+    docgen attrsets 'Attrset functions'
+    docgen strings 'String manipulation functions'
+    docgen trivial 'Miscellaneous functions'
+    docgen lists 'List manipulation functions'
+    docgen debug 'Debugging functions'
+    docgen options 'NixOS / nixpkgs option handling'
```
It looks like this is affected by https://github.com/NixOS/nixpkgs/pull/205643

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-01-19-documentation-team-meeting-notes-24/25537/1

ryantm committed:
nixpkgs-manual-mmdoc, nixos-manual-mmdoc: init

ryantm commented:
@NixOS/ofborg eval

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixpkgs-and-nixos-manual-enchancmenets-userscript/34203/5

ryantm commented:
I plan to merge this in one week.

06kellyjac commented on pkgs/tools/nix/nixpkgs-manual-mmdoc/toc.md:
```diff
@@ -0,0 +1,164 @@
+
+Version 23.11pre
+
+* [](#preface)
```
Why are all the links empty?

Are these all linked to content thats still TODO?
Maybe they should be commented out?

lf- replied:
I believe these have the title auto generated based on other instances of this pattern I've seen elsewhere.

SuperSandro2000 commented on pkgs/tools/nix/nixpkgs-manual-mmdoc-watch/default.nix:
```diff
@@ -0,0 +1,19 @@
+{ writeScriptBin, python3, entr } :
+let
+  np-build = writeScriptBin "np-build" ''
+    nix build .#nixpkgs-manual-mmdoc -L
```
This impurely depends on experimental-fearure nix-command which should be enabled with the command

Artturin replied:
```suggestion
     nix-build . -A nixpkgs-manual-mmdoc
```

lf- replied:
Also, due to lazy trees not yet existing, this is plain slower than `nix build -f  . nixpkgs-manual-mmdoc` (if we *are* using experimental syntax) or indeed `nix-build -A nixpkgs-manual-mmdoc`.

SuperSandro2000 commented on pkgs/tools/nix/nixpkgs-manual-mmdoc-watch/default.nix:
```diff
@@ -0,0 +1,19 @@
+  '';
+in
+writeScriptBin "nixpkgs-manual-mmdoc-watch" ''
+  killbg() {
```
Some formatting of the shell code would be nice

SuperSandro2000 commented on pkgs/tools/nix/nixpkgs-manual-mmdoc-watch/default.nix:
```diff
@@ -0,0 +1,19 @@
+  ${python3}/bin/python -m http.server --directory ./result &
+  pids+=($!)
+  trap exit SIGINT
+  while true; do find doc pkgs/tools/nix/nixpkgs-manual-mmdoc/ | ${entr}/bin/entr -cd ${np-build}/bin/np-build; done
```
Isn't entr Linux only and requires some setup? Last time I tried it, I couldn't get it to work.

SuperSandro2000 commented on pkgs/tools/nix/nixpkgs-manual-mmdoc-watch/default.nix:
```diff
@@ -0,0 +1,19 @@
+  ${python3}/bin/python -m http.server --directory ./result &
+  pids+=($!)
+  trap exit SIGINT
+  while true; do find doc pkgs/tools/nix/nixpkgs-manual-mmdoc/ | ${entr}/bin/entr -cd ${np-build}/bin/np-build; done
```
I am not sure if this results in a busy loop. Can't we use something like inotify-tools?

Artturin replied:
```suggestion
  while inotifywait -e modify -e create doc pkgs/tools/nix/nixpkgs-manual-mmdoc; do nix-build . -A nixpkgs-manual-mmdoc; done`
```

More uses if needed
https://github.com/search?q=lang%3Ash+%2Fwhile.%2Binotify%2F&type=code

SuperSandro2000 commented on pkgs/tools/nix/nixpkgs-manual-mmdoc/default.nix:
```diff
@@ -0,0 +1,17 @@
+
+  src = builtins.filterSource (path: type: type == "directory" || builtins.match ".*\.md" path == [] || builtins.match ".*\.dot" path == []) ../../../../doc;
+
+  phases = [ "buildPhase" ];
```
Please don't overwrite phases in packages and instead disable the phases which break the build our are really not wanted.

Artturin replied:
In this case probably only a `dontUnpack = true;` is needed

Artturin replied:
```suggestion
  dontUnpack = true;
```

SuperSandro2000 commented on pkgs/tools/nix/nixpkgs-manual-mmdoc/default.nix:
```diff
@@ -0,0 +1,17 @@
+    cp ${./toc.md} doc/toc.md
+    cp ${nixpkgs-manual-lib-docs}/*.md doc/functions/library/
+    ${mmdoc}/bin/mmdoc nixpkgs doc $out
+  '';
```
Some meta would be nice

Artturin commented on pkgs/tools/nix/nixpkgs-manual-mmdoc/default.nix:
```diff
@@ -0,0 +1,17 @@
+
+  phases = [ "buildPhase" ];
+
+  buildPhase = ''
```
The other phases don't seem necessary here so use `buildCommand`, no need to `dontUnpack`
```suggestion
  buildCommand = ''
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-01-04-documentation-team-meeting-notes-102/37754/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-01-25-documentation-team-meeting-notes-106/38792/1

infinisil commented:
This is being discussed a bit [in the docs Matrix room](https://matrix.to/#/!avYyleMexqjFHoqrME:nixos.org/$gZVNfFAnHxNtVGoLi-Mg_iesyv5uvyr1itgy0bvKLAA?via=nixos.org&via=matrix.org&via=nixos.dev)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/help-wanted-planning-for-summer-of-nix-2025/47556/2
